docs: add vap generation doc and demo (#3363)
open-policy-agent/gatekeeper: website/docs/validating-admission-policy.md

title: Integration with Kubernetes Validating Admission Policy
---

`Feature State`: Gatekeeper version v3.16 (alpha)

> ❗ This feature is alpha, subject to change (feedback is welcome!). It is disabled by default. To enable the feature,

> Install with Helm

Reviewer: do we want this to be a header?

Update the following parameters in values.yaml:
- Enable the K8s Native Validating driver to allow users to create CEL-based rules in addition to the OPA driver and rego rules (alpha feature). Default is `false`
    > enableK8sNativeValidation=true

> [optional] Install with Gatekeeper deployment

Reviewer: do we want this to be a header?

Update the following commandline flags:
- Set `--experimental-enable-k8s-native-validation=true`

## Description

This feature allows Gatekeeper to integrate with Kubernetes Validating Admission Policy based on [Common Expression Language (CEL)](https://github.com/google/cel-spec), a declarative, in-process admission control alternative to validating admission webhooks.

## Motivations

> The Validating Admission Policy feature (disabled by default) was introduced as an alpha feature to Kubernetes v1.26, beta in v1.28, CA in v1.30. Some of the benefits include:

Reviewer: ```suggestion
The Validating Admission Policy feature was introduced as an alpha feature to Kubernetes v1.26, beta in v1.28 (disabled by default), GA in v1.30 (enabled by default). Some of the benefits include:
```

The [Constraint Framework](https://github.com/open-policy-agent/frameworks/tree/master/constraint) is the library that underlies Gatekeeper. It provides the execution flow Gatekeeper uses to render a decision to the API server. It also provides abstractions that allow us to define constraint templates and constraints: Engine, Enforcement Points, and Targets.

Together with Gatekeeper and [gator CLI](gator.md), you can get admission, audit, and shift left validations for policies written in both CEL and Rego policy languages, even for clusters that do not support Validating Admission Policy feature yet. For simple policies, you may want admission requests to be handled by the K8s built-in Validating Admission Controller (only supports CEL) instead of the Gatekeeper admission webhook. 

To summary, these are potential options when running Gatekeeper:

> | Policy Language    | Enforcement Point  |

Reviewer: ```suggestion
| Policy Language(s)    | Enforcement Point  |
```

| ------------------ | ------------------ |

> | CEL, Rego          | Gatekeeper webhook |

Reviewer: ```suggestion
| CEL, Rego          | Gatekeeper validating webhook |
```

| CEL, Rego          | Gatekeeper Audit   |
| CEL, Rego          | Gator CLI          |
| CEL                | K8s built-in Validating Admission Controller (aka ValidatingAdmissionPolicy) |

> | Rego               | Gatekeeper webhook (referential policies) |

Reviewer: ```suggestion
| Rego               | Gatekeeper validating webhook (referential policies, external data) |
```

| Rego               | Gatekeeper Audit (referential policies) |

## Pre-requisites

- Requires minimum Gatekeeper v3.16.0
- Requires minimum Kubernetes v1.30, when the `Validating Admission Policy` feature GAed
- [optional] Kubernetes version v1.29, need to enable feature gate and runtime config as shown below: 

    ```yaml
    kind: Cluster
    apiVersion: kind.x-k8s.io/v1alpha4
    featureGates:
      ValidatingAdmissionPolicy: true
    runtimeConfig:
      admissionregistration.k8s.io/v1beta1: true
    ```
- Set `--experimental-enable-k8s-native-validation` in Gatekeeper deployments, or `enableK8sNativeValidation=true` if using Helm.

## Get started

> ### Install with Helm

Reviewer: ```suggestion
### Option 1: Install with Helm
```

> Update the following parameters in values.yaml:

Reviewer: ```suggestion
Update the `enableK8sNativeValidation` parameter in values.yaml or set during deployment
```

- Enable the K8s Native Validating driver to allow users to create CEL-based rules in addition to the OPA driver and rego rules (alpha feature). Default is `false`

>     > enableK8sNativeValidation=true

Reviewer: ```suggestion
```shell
helm install gatekeeper/gatekeeper --name-template=gatekeeper --namespace gatekeeper-system --create-namespace \
    *--set enableK8sNativeValidation=true*
```
```

Reviewer: there's ``` at the end, github doesn't like that as part of suggestion

> ### [optional] Install with Gatekeeper deployment

Reviewer: ```suggestion
### Option 2: Install with Gatekeeper deployment
```

> Update the following commandline flags:

Reviewer: ```suggestion
Edit the applicable deployments (controller-manager and audit), and update the following commandline flags:
```

The [Constraint Framework](https://github.com/open-policy-agent/frameworks/tree/master/constraint) is the library that underlies Gatekeeper. It provides the execution flow Gatekeeper uses to render a decision to the API server. It also provides abstractions that allow us to define constraint templates and constraints: Engine, Enforcement Points, and Targets.

Together with Gatekeeper and [gator CLI](gator.md), you can get admission, audit, and shift left validations for policies written in both CEL and Rego policy languages, even for clusters that do not support Validating Admission Policy feature yet. For simple policies, you may want admission requests to be handled by the K8s built-in Validating Admission Controller (only supports CEL) instead of the Gatekeeper admission webhook. 

> To summary, these are potential options when running Gatekeeper:

Author: Updated this, PTAL @sozercan @JaydipGabani @maxsmythe

The [Constraint Framework](https://github.com/open-policy-agent/frameworks/tree/master/constraint) is the library that underlies Gatekeeper. It provides the execution flow Gatekeeper uses to render a decision to the API server. It also provides abstractions that allow us to define constraint templates and constraints: Engine, Enforcement Points, and Targets.

Together with Gatekeeper and [gator CLI](gator.md), you can get admission, audit, and shift left validations for both CEL-based Validating Admission Policy and OPA Rego policies, even for clusters that do not support Validating Admission Policy feature yet. For some policies, you may want admission requests to be handled by the K8s Validating Admission Controller instead of the Gatekeeper admission webhook.

> 

Reviewer: can we summarize the 3 options GK will be able to do (rego, built-in cel and vap gen)

## Pre-requisites

> - Requires minimum Gatekeeper v3.16.0-beta.2

Reviewer: ```suggestion
- Requires minimum Gatekeeper v3.16.0
```

              messageExpression: '"missing required label, requires all of: " + variables.params.labels.map(entry, entry.key).join(", ")'
            - expression: '[object, oldObject].exists(obj, obj != null && !variables.params.labels.exists(entry, has(obj.metadata.labels) && entry.key in obj.metadata.labels && !string(obj.metadata.labels[entry.key]).matches(string(entry.allowedRegex))))'
              message: "regex mismatch"
        rego: |
          ...
```
With this new engine and source added to the constraint template, now Gatekeeper webhook, audit, and shift-left can validate resources with these new CEL-based rules. 

## Policy updates to generate Validating Admission Policy resources

For some policies, you may want admission requests to be handled by the K8s Validating Admission Controller instead of the Gatekeeper admission webhook. By default, Gatekeeper is configured to generate K8s Validating Admission Policy resources if the `gatekeeper.sh/use-vap` label is used. In the event K8s Validating Admission Controller fails open, then Gatekeeper admission webhook can act as a backup. Default value for this feature flag is `--vap-enforcement=GATEKEEPER_DEFAULT`. 

> Other allowed values are NONE: do not generate, `GATEKEEPER_DEFAULT`: do not generate unless label gatekeeper.sh/use-vap: yes is added to policy explicitly, `VAP_DEFAULT`: generate unless label gatekeeper.sh/use-vap: no is added to policy explicitly.

Reviewer: ```suggestion
Other allowed values are `NONE`: do not generate, `GATEKEEPER_DEFAULT`: do not generate unless label `gatekeeper.sh/use-vap: yes` is added to policy explicitly, `VAP_DEFAULT`: generate unless label `gatekeeper.sh/use-vap: no` is added to policy explicitly.
```
